python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx

```diff
@@ -0,0 +1,371 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# cython: language_level=3
+
+from libc.stdint cimport uint32_t
+from libcpp cimport bool
+from libcpp.string cimport string
+
+from cuvs.common.exceptions import check_cuvs
+from cuvs.common.resources import auto_sync_resources
+
+from cuvs.common cimport cydlpack
+
+import numpy as np
+
+from cuvs.distance import DISTANCE_TYPES
+
+from cuvs.neighbors.cagra cimport cagra
+
+import os
+import uuid
+
+from pylibraft.common import auto_convert_output
+from pylibraft.common.cai_wrapper import wrap_array
+from pylibraft.common.interruptible import cuda_interruptible
+from pylibraft.neighbors.common import _check_input_array
+
+
+cdef class SearchParams:
+    """
+    HNSW search parameters
+
+    Parameters
+    ----------
+    ef: int, default = 200
+        Maximum number of candidate list size used during search.
+    num_threads: int, default = 0
+        Number of CPU threads used to increase search parallelism.
+        When set to 0, the number of threads is automatically determined.
+    """
+
+    cdef cuvsHnswSearchParams params
+
+    def __init__(self, *,
+                 ef=200,
+                 num_threads=0):
+        self.params.ef = ef
+        self.params.numThreads = num_threads
+
+    def __repr__(self):
+        attr_str = [attr + "=" + str(getattr(self, attr))
+                    for attr in [
+                        "ef", "num_threads"]]
+        return "SearchParams(type=HNSW, " + (", ".join(attr_str)) + ")"
+
+    @property
+    def ef(self):
+        return self.params.ef
+
+    @property
+    def num_threads(self):
+        return self.params.numThreads
+
+
+cdef class Index:
+    """
+    HNSW index object. This object stores the trained HNSW index state
+    which can be used to perform nearest neighbors searches.
+    """
+
+    cdef cuvsHnswIndex_t index
+    cdef bool trained
+
+    def __cinit__(self):
+        self.trained = False
+        check_cuvs(cuvsHnswIndexCreate(&self.index))
+
+    def __dealloc__(self):
+        if self.index is not NULL:
+            check_cuvs(cuvsHnswIndexDestroy(self.index))
+
+    @property
+    def trained(self):
+        return self.trained
+
+    def __repr__(self):
+        # todo(dgd): update repr as we expose data through C API
+        attr_str = []
+        return "Index(type=HNSW, metric=L2" + (", ".join(attr_str)) + ")"
+
+
+@auto_sync_resources
+def save(filename, cagra.Index index, resources=None):
+    """
+    Saves the CAGRA index to a file as an hnswlib index.
+    The saved index is immutable and can only be searched by the hnswlib
+    wrapper in cuVS, as the format is not compatible with the original
+    hnswlib.
+
+    Saving / loading the index is experimental. The serialization format is
```

Comment:
Please include the note here about how even though this format closely matches hnswlib, it is only intended to be used by cuVS hnsw apis.

Reply:
That note is already present right above
```
    The saved index is immutable and can only be searched by the hnswlib
    wrapper in cuVS, as the format is not compatible with the original
    hnswlib.
```